flink-contrib/flink-timerserivce-rocksdb/src/main/java/org/apache/flink/contrib/streaming/api/operators/RocksDBInternalTimerService.java

```diff
@@ -0,0 +1,797 @@
+			eventTimeColumnFamilyHandle = db.createColumnFamily(eventTimeColumnFamilyDescriptor);
+			processingTimeColumnFamilyHandle = db.createColumnFamily(processingTimeColumnFamilyDescriptor);
+		} catch (RocksDBException e) {
+			throw new RuntimeException("Error while creating the column families.", e);
+		}
+
+		this.processingTimeHeap = new PersistentTimerHeap(numPartitions, processingTimeColumnFamilyHandle);
+		this.eventTimeHeap = new PersistentTimerHeap(numPartitions, eventTimeColumnFamilyHandle);
+	}
+
+	// ------------------------------------------------------------------------
+	//  InternalTimerService Implementation
+	// ------------------------------------------------------------------------
+	
+	@Override
+	public void start() {
+		// rebuild the heaps
+		eventTimeHeap.initialize();
+		processingTimeHeap.initialize();
+		
+		// register the processing timer with the minimum timestamp
+		Tuple4<Integer, Long, K, N> headProcessingTimer = processingTimeHeap.top();
+		if (headProcessingTimer != null) {
+			nextTimer = processingTimeService.registerTimer(headProcessingTimer.f1, this);
+		}
+	}
+
+	@Override
+	public void close() {
+		if (db != null) {
+			db.close();
+		}
+		
+		if (dbPath != null) {
+			try {
+				FileSystem fileSystem = dbPath.getFileSystem();
+				if (fileSystem.exists(dbPath)) {
+					fileSystem.delete(dbPath, true);
+				}
+			} catch (IOException e) {
+				throw new RuntimeException("Error while cleaning directory for rocksdb timer service.", e);
+			}
+		}
+	}
+
+	@Override
+	public void onEventTime(long timestamp) throws Exception {
+		List<Tuple4<Integer, Long, K, N>> timers = eventTimeHeap.peek(timestamp);
+		for (Tuple4<Integer, Long, K, N> timer : timers) {
+			keyContext.setCurrentKey(timer.f2);
+			triggerTarget.onEventTime(new InternalTimer<>(timer.f1, timer.f2, timer.f3));
+		}
+	}
+
+	@Override
+	public void onProcessingTime(long timestamp) throws Exception {
+		nextTimer = null;
+
+		List<Tuple4<Integer, Long, K, N>> timers = processingTimeHeap.peek(timestamp);
+		for (Tuple4<Integer, Long, K, N> timer : timers) {
+			keyContext.setCurrentKey(timer.f2);
+			triggerTarget.onProcessingTime(new InternalTimer<>(timer.f1, timer.f2, timer.f3));
+		}
+
+		if (nextTimer == null) {
+			Tuple4<Integer, Long, K, N> headTimer = processingTimeHeap.top();
+			if (headTimer != null) {
+				nextTimer = processingTimeService.registerTimer(headTimer.f1, this);
+			}
+		}
+	}
+
+	@SuppressWarnings("unchecked")
+	@Override
+	public void registerProcessingTimeTimer(N namespace, long time) {
+		boolean isNewHead = processingTimeHeap.add((K)keyContext.getCurrentKey(), namespace, time);
+
+		if (isNewHead) {
+			if (nextTimer != null) {
+				nextTimer.cancel(false);
+			}
+
+			Tuple4<Integer, Long, K, N> newHeadTimer = processingTimeHeap.top();
+			if (newHeadTimer == null || newHeadTimer.f1 != time) {
+				throw new IllegalStateException();
```

Comment:
The exception should contain a message about what went wrong. This should also be fixed in similar cases in this class.